src/Common/src/System/Net/Http/TlsCertificateExtensions.cs

```diff
@@ -0,0 +1,100 @@
+
+namespace System.Net.Http
+{
+    /// <summary>
+    /// Defines Extension methods which are meant to be re-used across WinHttp & UnixHttp Handlers
+    /// </summary>
+    internal static class TLSCertificateExtensions
+    {
+        private const string ClientCertificateOid = "1.3.6.1.5.5.7.3.2";
+
+        /// <summary>
+        ///   returns true if the X509 Certificate can be used  as SSL Client Certificate.
+        /// </summary>
+        internal static bool CanAuthenticateClient(this X509Certificate2 cert)
+        {
+            if (cert == null || cert.Extensions == null)
+            {
+                return false;
+            }
+
+            foreach(X509Extension extension in cert.Extensions)
+            {
+                X509EnhancedKeyUsageExtension enhancedUsageExt = extension as X509EnhancedKeyUsageExtension;
+                if (enhancedUsageExt != null)
+                {
+                    foreach(Oid oid in enhancedUsageExt.EnhancedKeyUsages)
+                    {
+                        if (string.Equals(ClientCertificateOid, oid.Value))
+                        {
+                            return true;
+                        }
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        ///   Returns true if the given X509 Certifcate can be used as a digital signature
+        /// </summary>
+        internal static bool IsDigitalSignature(this X509Certificate2 cert)
+        {
+            if (cert == null || cert.Extensions == null)
+            {
+                return false;
+            }
+
+            foreach(X509Extension extension in cert.Extensions)
+            {
+                X509EnhancedKeyUsageExtension enhancedUsageExt = extension as X509EnhancedKeyUsageExtension;
+                if (enhancedUsageExt != null)
+                {
+                    foreach(Oid oid in enhancedUsageExt.EnhancedKeyUsages)
+                    {
+                        if (string.Equals(ClientCertificateOid, oid.Value))
+                        {
+                            return true;
+                        }
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        
+        /// <summary>
+        ///   Returns a new collection containing valid client certificates from the given X509Certificate2Collection
+        /// </summary>
+        internal static X509Certificate2Collection FindTlsClientCertificates(this X509Certificate2Collection certificates)
+        {
+            X509Certificate2Collection clientCertificates = new X509Certificate2Collection();
+            if (certificates == null)
+            {
+                return clientCertificates;
+            }
+
+            DateTime now = DateTime.Now;
+            foreach(X509Certificate2 cert in certificates)
+            {
+                if (cert.HasPrivateKey &&  now > cert.NotBefore &&   now < cert.NotAfter)
```

Comment:
Well, it would require that the cert chain be buildable, and trusted.  Which is weird, why would you care if the local machine trusts the client authentication certificate it's going to use?